x-pack/test/functional/apps/maps/add_layer_panel.js

```diff
@@ -0,0 +1,89 @@
+  const PageObjects = getPageObjects(['maps', 'common']);
+
+  describe('Add layer panel', () => {
+    before(async () => {
+      await PageObjects.maps.openNewMap();
+    });
+
+    beforeEach(async () => {
+      await PageObjects.maps.clickAddLayer();
+    });
+
+    afterEach(async () => {
+      await PageObjects.maps.cancelLayerAdd();
+    });
+
+    it('should open on clicking "Add layer"', async () => {
+      // Verify panel page element is open
+      const panelOpen = await PageObjects.maps.isLayerAddPanelOpen();
+      expect(panelOpen).to.be(true);
+    });
+
+    it('should close on clicking "Cancel"', async () => {
+      // Verify panel page element is open
+      let panelOpen = await PageObjects.maps.isLayerAddPanelOpen();
+      expect(panelOpen).to.be(true);
+      // Click cancel
+      await PageObjects.maps.cancelLayerAdd();
+      // Verify panel isn't open
+      panelOpen = await PageObjects.maps.isLayerAddPanelOpen();
+      expect(panelOpen).to.be(false);
+    });
+
+    it('should close & remove layer on clicking "Cancel" after selecting layer',
+      async () => {
+        // Verify panel page element is open
+        let panelOpen = await PageObjects.maps.isLayerAddPanelOpen();
+        expect(panelOpen).to.be(true);
+        // Select source
+        await PageObjects.maps.selectVectorSource();
+        // Select layer
+        const vectorLayer = await PageObjects.maps.selectVectorLayer();
+        // Confirm layer added
+        await PageObjects.common.sleep(1000);
+        let vectorLayerExists = await PageObjects.maps.doesLayerExist(vectorLayer);
+        expect(vectorLayerExists).to.be(true);
+        // Click cancel
+        await PageObjects.maps.cancelLayerAdd();
+        // Verify panel isn't open
+        panelOpen = await PageObjects.maps.isLayerAddPanelOpen();
+        expect(panelOpen).to.be(false);
+        // Verify layer has been removed
+        await PageObjects.common.sleep(500);
+        vectorLayerExists = await PageObjects.maps.doesLayerExist(vectorLayer);
+        expect(vectorLayerExists).to.be(false);
+      });
+
+    it('should close and remove layer on map save', async () => {
+      // Verify panel page element is open
+      let panelOpen = await PageObjects.maps.isLayerAddPanelOpen();
+      expect(panelOpen).to.be(true);
+      // Select source
+      await PageObjects.maps.selectVectorSource();
+      // Select layer
+      const vectorLayer = await PageObjects.maps.selectVectorLayer();
+      // Confirm layer added
+      await PageObjects.common.sleep(1000);
+      let vectorLayerExists = await PageObjects.maps.doesLayerExist(vectorLayer);
+      expect(vectorLayerExists).to.be(true);
+      // Click save
+      await PageObjects.maps.saveMap('Mappishness');
+      // Verify panel isn't open
+      panelOpen = await PageObjects.maps.isLayerAddPanelOpen();
+      expect(panelOpen).to.be(false);
+      // Verify layer has been removed
+      await PageObjects.common.sleep(500);
```

Comment:
Instead of sleep, wrap expect in retry. Sleeps lead to flaky tests because they mask timing issues